### PR TITLE
Fix keepZeroFacets on disjunctive facet search

### PR DESCRIPTION
**Pull request by bidoubiwa**
Since introducing the multi-index search here #888 the setting `keepZeroFacets` was broken.

This PR introduces the fix to this issue

https://user-images.githubusercontent.com/33010418/207385436-e49ded20-e76b-4974-86de-904cfeecfbd7.mp4

What was done in this pr:

- keepZeroFacet management has been moved to the response adapter instead of the request adapter
- Filtering methods and filtering tests were removed as the new facet distribution management has completely changed
- `defaultFacetDistribution` is renamed `initialFacetDistribution` and contains the distribution of an index without any other requests
- Initialization of facet distribution was moved from client to its own file

Playgrounds:
- New SingleMovieIndex playground to create a new environment for testing
- Updated `SingleIndex` with keepZeroFacets
- Updated `MultiIndex` with keepZeroFacets

### CODE REVIEW

File: src/adapter/search-request-adapter/search-resolver.ts [hunk 1]
```diff
@@ -42,16 +41,6 @@ export function SearchResolver(
...
-      if (searchContext.keepZeroFacets) {
-        const cachedFacets = extractFacets(searchContext, searchParams)
-
-        // Add missing facets back into facetDistribution
-        searchResponse.facetDistribution = addMissingFacets(
-          cachedFacets,
-          searchResponse.facetDistribution
-        )
-      }
-
```
Comment (bidoubiwa) [inline src/adapter/search-request-adapter/search-resolver.ts:54]: Moved as it should be in the response adapter

File: src/client/instant-meilisearch-client.ts [hunk 2]
```diff
@@ -72,27 +72,22 @@ export function instantMeiliSearch(
...
-          // Cache first facets distribution of the instantMeilisearch instance
-          // Needed to add in the facetDistribution the fields that were not returned
-          // When the user sets `keepZeroFacets` to true.
-          if (defaultFacetDistribution === undefined) {
-            defaultFacetDistribution = await cacheFirstFacetDistribution(
-              searchResolver,
-              searchContext
-            )
-            searchContext.defaultFacetDistribution = defaultFacetDistribution
-          }
```
Comment (bidoubiwa) [inline src/client/instant-meilisearch-client.ts:95]: removed to its own file `init-facet-distribution.ts`

File: src/client/instant-meilisearch-client.ts [hunk 3]
```diff
@@ -55,7 +55,7 @@ export function instantMeiliSearch(
   // create search resolver with included cache
   const searchResolver = SearchResolver(meilisearchClient, searchCache)
 
-  let defaultFacetDistribution: FacetDistribution
+  let initialFacetDistribution: Record<string, FacetDistribution>
```
Comment (bidoubiwa) [inline src/client/instant-meilisearch-client.ts:58]: Changed name to better understand what it contains

File: src/client/instant-meilisearch-client.ts [hunk 4]
```diff
@@ -103,11 +98,13 @@ export function instantMeiliSearch(
           // Adapt the Meilisearch response to a compliant instantsearch.js response
           const adaptedSearchResponse = adaptSearchResponse<T>(
             searchResponse,
-            searchContext
+            searchContext,
+            initialFacetDistribution[searchRequest.indexName]
```
Comment (bidoubiwa) [inline src/client/instant-meilisearch-client.ts:101]: removed `initialFacetDistribution` from the `searchContext` as it is not correlated to the current search request but as value for the whole instance

File: src/contexts/search-context.ts [hunk 5]
```diff
@@ -33,7 +31,6 @@ export function createSearchContext(
     sort: sortByArray.join(':') || '',
     indexUid,
     pagination: paginationState,
-    defaultFacetDistribution: defaultFacetDistribution || {},
```
Comment (bidoubiwa) [inline src/contexts/search-context.ts:36]: removed `initialFacetDistribution` from the `searchContext` as it is not correlated to the current search request but as value for the whole instance.

File: src/types/types.ts [hunk 6]
```diff
@@ -73,9 +70,10 @@ export type InstantSearchPagination = {
   nbPages: number
 }
 
+export type Facets = string | string[] | undefined
+
 export type SearchContext = Omit<InstantSearchParams, 'insideBoundingBox'> &
   InstantSearchParams & {
-    defaultFacetDistribution: FacetDistribution
```
Comment (bidoubiwa) [inline src/types/types.ts:78]: removed `defaultFacetDistribution ` from the `searchContext` as it is not correlated to the current search request but as value for the whole instance.

File: src/types/types.ts [hunk 7]
```diff
@@ -73,9 +70,10 @@ export type InstantSearchPagination = {
   nbPages: number
 }
 
+export type Facets = string | string[] | undefined
```
Comment (bidoubiwa) [inline src/types/types.ts:73]: Creation of own `Facets` type as the one from instant-search does not represent the reality.

`instant-search` may provide only a `string` as a `Facets` but its type states that it only accepts `string[] | undefined`

File: src/adapter/search-request-adapter/search-resolver.ts [hunk 8]
```diff
@@ -5,7 +5,6 @@ import {
   SearchCacheInterface,
   MeiliSearchParams,
 } from '../../types'
-import { addMissingFacets, extractFacets } from './filters'
```
Comment (bidoubiwa) [inline src/adapter/search-request-adapter/search-resolver.ts:8]: removed as facet managements has changed and moved to the response adapter

File: src/adapter/search-response-adapter/facet-distribution-adapter.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  else if (typeof facets === 'string') return [facets]
+  return facets
+}
+
+// Fills the missing facetValue in the current facet distribution if `keepZeroFacet` is true
+// using the initial facet distribution.
+// Ex:
+// Initial distribution: { genres: { horror: 10, comedy: 4 } }
+// Current distribution: { genres: { horror: 3 }}
+// returned distribution: { genres: { horror: 3, comedy: 0 }}
```
Comment (brunoocasali) [inline src/adapter/search-response-adapter/facet-distribution-adapter.ts:16]:
```suggestion
// Fills the missing facetValue in the current facet distribution if `keepZeroFacet` is true
// using the initial facet distribution. Ex:
//
// Initial distribution: { genres: { horror: 10, comedy: 4 } }
// Current distribution: { genres: { horror: 3 }}
// Returned distribution: { genres: { horror: 3, comedy: 0 }}
```

File: src/adapter/search-response-adapter/facet-distribution-adapter.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        filledDistribution[facet] = currentFacetDistribution[facet] || {}
+      }
+      if (!filledDistribution[facet][facetValue]) {
+        filledDistribution[facet][facetValue] = 0
+      } else {
+        filledDistribution[facet][facetValue] =
+          currentFacetDistribution[facet][facetValue]
+      }
+    }
+  }
```
Comment (brunoocasali) [inline src/adapter/search-response-adapter/facet-distribution-adapter.ts:38]: I didn't find a way to make it faster, but you can make it look better by removing the word "distribution" :)

**Review by brunoocasali — APPROVED** [review]
I left some comments that can be done in a new PR maybe, I don't want to block the merge :)

LGTM!

**Comment by bidoubiwa** [comment]
bors merge

**Pull request merged**